File: packages/nerva-js/src/runtime/streaming.ts
/**
 * Streaming runtime wrapper — push structured token chunks to ctx.stream
 * as the LLM produces them.
 *
 * Wraps any AgentRuntime and intercepts output to push structured chunks
 * to the context's stream sink.
 *
 * @module runtime/streaming
 */

import type { ExecContext } from "../context.js";
import type { AgentInput, AgentResult, AgentRuntime } from "./index.js";
import { AgentStatus } from "./index.js";

// ---------------------------------------------------------------------------
// StreamChunkType
// ---------------------------------------------------------------------------

/**
 * Kind of streaming chunk pushed to the sink.
 *
 * - `token` — an incremental text token from the LLM.
 * - `progress` — a progress indicator (e.g. percentage or status message).
 * - `complete` — final chunk signalling the stream is done.
 * - `error` — an error occurred during streaming.
 */
export type StreamChunkType = "token" | "progress" | "complete" | "error";

// ---------------------------------------------------------------------------
// StreamChunk
// ---------------------------------------------------------------------------

/**
 * A single streaming chunk pushed to the context's stream sink.
 */
export interface StreamChunk {
  /** The kind of chunk. */
  readonly type: StreamChunkType;
  /** Text payload of the chunk. */
  readonly content: string;
  /** Unix timestamp when the chunk was created. */
  readonly timestamp: number;
}

// ---------------------------------------------------------------------------
// StreamingRuntime
// ---------------------------------------------------------------------------

/**
 * Wraps any AgentRuntime to push structured stream chunks to `ctx.stream`.
 *
 * When a handler is invoked, this wrapper delegates to the inner runtime
 * and pushes a COMPLETE chunk with the full output. If the handler fails,
 * an ERROR chunk is pushed instead.
 */
export class StreamingRuntime implements AgentRuntime {
  private readonly _inner: AgentRuntime;

  /**
   * @param inner - The underlying runtime to delegate execution to.
   */
  constructor(inner: AgentRuntime) {
    this._inner = inner;
  }

  /**
   * Invoke a handler and push structured chunks to `ctx.stream`.
   *
   * @param handler - Handler name to invoke.
   * @param input - Structured input for the handler.
   * @param ctx - Execution context with optional stream sink.
   * @returns AgentResult from the underlying runtime.
   */
  async invoke(
    handler: string,
    input: AgentInput,
    ctx: ExecContext,
  ): Promise<AgentResult> {
    const result = await this._inner.invoke(handler, input, ctx);
    await pushResultChunk(result, ctx);
    return result;
  }

  /**
   * Run handlers in sequence, pushing a final chunk for the last result.
   *
   * @param handlers - Ordered list of handler names.
   * @param input - Initial input for the first handler.
   * @param ctx - Execution context shared across the chain.
   * @returns AgentResult from the last successfully executed handler.
   */
  async invokeChain(
    handlers: string[],
    input: AgentInput,
    ctx: ExecContext,
  ): Promise<AgentResult> {
    const result = await this._inner.invokeChain(handlers, input, ctx);
    await pushResultChunk(result, ctx);
    return result;
  }

  /**
   * Delegate to another handler, pushing a chunk for the result.
   *
   * @param handler - Handler name to delegate to.
   * @param input - Input for the delegated handler.
   * @param parentCtx - Parent's execution context.
   * @returns AgentResult from the delegated handler.
   */
  async delegate(
    handler: string,
    input: AgentInput,
    parentCtx: ExecContext,
  ): Promise<AgentResult> {
    const result = await this._inner.delegate(handler, input, parentCtx);
    await pushResultChunk(result, parentCtx);
    return result;
  }
}

// ---------------------------------------------------------------------------
// Pure helpers
// ---------------------------------------------------------------------------

/**
 * Create a StreamChunk with the current timestamp.
 *
 * @param type - Kind of chunk.
 * @param content - Text payload.
 * @returns A StreamChunk object.
 */
export function buildChunk(type: StreamChunkType, content: string): StreamChunk {
  return { type, content, timestamp: Date.now() / 1000 };
}

/**
 * Serialize a StreamChunk to a JSON string for the stream sink.
 *
 * @param chunk - The chunk to serialize.
 * @returns JSON string representation.
 */
export function serializeChunk(chunk: StreamChunk): string {
  return JSON.stringify({
    type: chunk.type,
    content: chunk.content,
    timestamp: chunk.timestamp,
  });
}

/**
 * Push a structured chunk to the stream sink based on the result status.
 *
 * No-ops if the context has no stream attached.
 *
 * @param result - The agent result to convert to a chunk.
 * @param ctx - Execution context with optional stream sink.
 */
async function pushResultChunk(
  result: AgentResult,
  ctx: ExecContext,
): Promise<void> {
  if (ctx.stream === null) return;

  const chunk =
    result.status === AgentStatus.SUCCESS
      ? buildChunk("complete", result.output)
      : buildChunk("error", result.error ?? `handler failed with status ${result.status}`);

  await ctx.stream.push(serializeChunk(chunk));
}

File: packages/nerva-py/nerva/runtime/streaming.py
"""Streaming runtime wrapper — push tokens to ctx.stream as the LLM produces them.

Wraps any AgentRuntime and intercepts output to push structured chunks
to the context's stream sink. Supports two modes:

1. Handler returns an async generator -- each yielded value is pushed as a token chunk.
2. Handler returns a string -- the entire result is pushed as a single complete chunk.

Each chunk is a JSON-serialisable dict with type, content, and timestamp.
"""

from __future__ import annotations

import time
from dataclasses import dataclass
from enum import StrEnum
from typing import TYPE_CHECKING

from nerva.runtime import AgentInput, AgentResult, AgentStatus

if TYPE_CHECKING:
    from nerva.context import ExecContext
    from nerva.runtime import AgentRuntime

__all__ = [
    "StreamChunkType",
    "StreamChunk",
    "StreamingRuntime",
]


# ---------------------------------------------------------------------------
# Constants & types
# ---------------------------------------------------------------------------


class StreamChunkType(StrEnum):
    """Kind of streaming chunk pushed to the sink.

    Members:
        TOKEN: An incremental text token from the LLM.
        PROGRESS: A progress indicator (e.g. percentage or status message).
        COMPLETE: Final chunk signalling the stream is done.
        ERROR: An error occurred during streaming.
    """

    TOKEN = "token"
    PROGRESS = "progress"
    COMPLETE = "complete"
    ERROR = "error"


@dataclass(frozen=True)
class StreamChunk:
    """A single streaming chunk pushed to the context's stream sink.

    Attributes:
        type: The kind of chunk.
        content: Text payload of the chunk.
        timestamp: Unix timestamp when the chunk was created.
    """

    type: StreamChunkType
    content: str
    timestamp: float


# ---------------------------------------------------------------------------
# StreamingRuntime
# ---------------------------------------------------------------------------


class StreamingRuntime:
    """Wraps any AgentRuntime to push structured stream chunks to ``ctx.stream``.

    When a handler is invoked, this wrapper delegates to the inner runtime
    and pushes a COMPLETE chunk with the full output. If the handler fails,
    an ERROR chunk is pushed instead.

    For async-generator handlers (streaming at the InProcessRuntime level),
    the inner runtime already pushes raw text to ``ctx.stream``. This wrapper
    adds structured framing on top.

    Args:
        inner: The underlying runtime to delegate execution to.
    """

    def __init__(self, inner: AgentRuntime) -> None:
        self._inner = inner

    async def invoke(
        self, handler: str, input: AgentInput, ctx: ExecContext
    ) -> AgentResult:
        """Invoke a handler and push structured chunks to ``ctx.stream``.

        Args:
            handler: Handler name to invoke.
            input: Structured input for the handler.
            ctx: Execution context with optional stream sink.

        Returns:
            AgentResult from the underlying runtime.
        """
        result = await self._inner.invoke(handler, input, ctx)
        await _push_result_chunk(result, ctx)
        return result

    async def invoke_chain(
        self, handlers: list[str], input: AgentInput, ctx: ExecContext
    ) -> AgentResult:
        """Run handlers in sequence, pushing a final chunk for the last result.

        Args:
            handlers: Ordered list of handler names.
            input: Initial input for the first handler.
            ctx: Execution context shared across the chain.

        Returns:
            AgentResult from the last successfully executed handler.
        """
        result = await self._inner.invoke_chain(handlers, input, ctx)
        await _push_result_chunk(result, ctx)
        return result

    async def delegate(
        self, handler: str, input: AgentInput, parent_ctx: ExecContext
    ) -> AgentResult:
        """Delegate to another handler, pushing a chunk for the result.

        Args:
            handler: Handler name to delegate to.
            input: Input for the delegated handler.
            parent_ctx: Parent's execution context.

        Returns:
            AgentResult from the delegated handler.
        """
        result = await self._inner.delegate(handler, input, parent_ctx)
        await _push_result_chunk(result, parent_ctx)
        return result


# ---------------------------------------------------------------------------
# Pure helpers
# ---------------------------------------------------------------------------


def build_chunk(chunk_type: StreamChunkType, content: str) -> StreamChunk:
    """Create a StreamChunk with the current timestamp.

    Args:
        chunk_type: Kind of chunk.
        content: Text payload.

    Returns:
        A frozen StreamChunk instance.
    """
    return StreamChunk(type=chunk_type, content=content, timestamp=time.time())


def serialize_chunk(chunk: StreamChunk) -> str:
    """Serialize a StreamChunk to a JSON-compatible string for the stream sink.

    Args:
        chunk: The chunk to serialize.

    Returns:
        JSON string representation of the chunk.
    """
    import json

    return json.dumps({
        "type": chunk.type.value,
        "content": chunk.content,
        "timestamp": chunk.timestamp,
    })


async def _push_result_chunk(result: AgentResult, ctx: ExecContext) -> None:
    """Push a structured chunk to the stream sink based on the result status.

    Args:
        result: The agent result to convert to a chunk.
        ctx: Execution context with optional stream sink.
    """
    if ctx.stream is None:
        return

    if result.status == AgentStatus.SUCCESS:
        chunk = build_chunk(StreamChunkType.COMPLETE, result.output)
    else:
        error_msg = result.error or f"handler failed with status {result.status}"
        chunk = build_chunk(StreamChunkType.ERROR, error_msg)

    await ctx.stream.push(serialize_chunk(chunk))
